OpenNefia.Content/UI/Hud/Widgets/HudBarBaseWidget.cs

```diff
@@ -0,0 +1,109 @@
+﻿using OpenNefia.Content.UI;
+using OpenNefia.Content.UI.Element;
+using OpenNefia.Content.Prototypes;
+using OpenNefia.Core.Maths;
+using OpenNefia.Core.Rendering;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenNefia.Core.IoC;
+using OpenNefia.Core.GameObjects;
+using OpenNefia.Content.Skills;
+using OpenNefia.Core.Game;
+
+namespace OpenNefia.Content.Hud
+{
+    public abstract class HudBarBaseWidget : BaseHudWidget
+    {
+
+        protected abstract string BarText { get; set; }
+        protected abstract IAssetInstance BarAsset { get; set; }
+        protected abstract float BarRatio { get; set; }
+        protected virtual Vector2i BarSize { get; } = new(83, 7);
+        protected virtual Vector2 BarOffset { get; } = new(17, 5);
+
+        private UiHelpers.UiBarDrawableState BarState = default!;
+        private IAssetInstance BarBGAsset;
+        private IUiText UiText;
+
+        public HudBarBaseWidget()
+        {
+            UiText = new UiTextOutlined(UiFonts.HUDBarText);
+            BarBGAsset = Assets.Get(Protos.Asset.HpBarFrame);
+        }
+
+        public override void SetPosition(int x, int y)
+        {
+            base.SetPosition(x, y);
+            UiText.SetPosition(x + 20, y - 7);
+        }
+
+        protected void SetState()
```

Comment:
Maybe `RefreshBarState()` instead, to make it clear this is a bar-related method.

Reply:
changed in 88260a8